File: tutorials/mpi/examples/mpi_helloBsend.c
/******************************************************************************
* FILE: mpi_helloBsend.c
* DESCRIPTION:
*   MPI tutorial example code: Simple hello world program that uses blocking
*   send/receive routines.
* AUTHOR: Blaise Barney
* LAST REVISED: 06/08/15
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#define  MASTER		0

int main (int argc, char *argv[])
{
int  numtasks, taskid, len, partner, message;
char hostname[MPI_MAX_PROCESSOR_NAME];
MPI_Status status;

MPI_Init(&argc, &argv);
MPI_Comm_rank(MPI_COMM_WORLD, &taskid);
MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

/* need an even number of tasks  */
if (numtasks % 2 != 0) {
   if (taskid == MASTER) 
      printf("Quitting. Need an even number of tasks: numtasks=%d\n", numtasks);
   } 

else {
   if (taskid == MASTER) 
      printf("MASTER: Number of MPI tasks is: %d\n",numtasks);

   MPI_Get_processor_name(hostname, &len);
   printf ("Hello from task %d on %s!\n", taskid, hostname);

   /* determine partner and then send/receive with partner */
   if (taskid < numtasks/2) {
     partner = numtasks/2 + taskid;
     MPI_Send(&taskid, 1, MPI_INT, partner, 1, MPI_COMM_WORLD);
     MPI_Recv(&message, 1, MPI_INT, partner, 1, MPI_COMM_WORLD, &status);
     }
   else if (taskid >= numtasks/2) {
     partner = taskid - numtasks/2;
     MPI_Recv(&message, 1, MPI_INT, partner, 1, MPI_COMM_WORLD, &status);
     MPI_Send(&taskid, 1, MPI_INT, partner, 1, MPI_COMM_WORLD);
     }

   /* print partner info and exit*/
   printf("Task %d is partner with %d\n",taskid,message);
   }

MPI_Finalize();

}

File: mpi/examples/mpi_helloBsend.f
C ****************************************************************************
C FILE: mpi_helloBsend.f
C DESCRIPTION:
C   MPI tutorial example code: Simple hello world program that uses
C   blocking send/receive routines
C AUTHOR: Blaise Barney
C LAST REVISED: 06/08/15
C ****************************************************************************
      program hello
      include 'mpif.h'
      parameter (MASTER = 0)

      integer numtasks, taskid, len, rem, ierr
      integer partner, message, status(MPI_STATUS_SIZE)
      character(MPI_MAX_PROCESSOR_NAME) hostname

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, taskid, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

C need an even number of tasks
      rem = mod(numtasks,2)
      if (rem .ne. 0) then
        if (taskid .eq. MASTER) then
          write(*,10) numtasks
          endif
         
      else
        call MPI_GET_PROCESSOR_NAME(hostname, len, ierr)
        write(*,20) taskid, hostname
        if (taskid .eq. MASTER) then
          write(*,30) numtasks
        endif

C determine partner and then send/receive with partner
        if (taskid .lt. numtasks/2) then
          partner = numtasks/2 + taskid
          call MPI_SEND(taskid, 1, MPI_INTEGER, partner, 1, 
     &                  MPI_COMM_WORLD, ierr)
          call MPI_RECV(message, 1, MPI_INTEGER, partner, 1, 
     &                  MPI_COMM_WORLD, status, ierr)

        else if (taskid .ge. numtasks/2) then
          partner = taskid - numtasks/2
          call MPI_RECV(message, 1, MPI_INTEGER, partner, 1,
     &                  MPI_COMM_WORLD, status, ierr)
          call MPI_SEND(taskid, 1, MPI_INTEGER, partner, 1,
     &                MPI_COMM_WORLD, ierr)
        endif
  
C print partner info and exit
        write(*,40) taskid, message

      endif

      call MPI_FINALIZE(ierr)
  

  10  format('Quitting. Need an even number of tasks: numtasks=', I2)
  20  format('Hello from task ',I2,' on ',A48)
  30  format('MASTER: Number of MPI tasks is: ',I2)
  40  format('Task ',I2,' is partner with ',I2)

      end
